realm/realm-library/src/main/java/io/realm/sync/permissions/Permission.java

```diff
@@ -0,0 +1,562 @@
+
+    /**
+     * Creates a {@link Permission} object in a fluid manner.
+     */
+    public static class Builder {
+        private Role role;
+        private boolean canRead = false;
+        private boolean canUpdate = false;
+        private boolean canDelete = false;
+        private boolean canSetPermissions = false;
+        private boolean canQuery = false;
+        private boolean canCreate = false;
+        private boolean canModifySchema = false;
+
+        /**
+         * Creates the builder. The default state is that no privileges are enabled.
+         *
+         * @param role {@link Role} for which these privileges apply.
+         */
+        public Builder(Role role) {
+            this.role = role;
+        }
+
+        /**
+         * Enables all privileges.
+         */
+        public Builder allPrivileges() {
+            canRead = false;
+            canUpdate = false;
+            canDelete = false;
+            canSetPermissions = false;
+            canQuery = false;
+            canCreate = false;
+            canModifySchema = false;
+            return this;
+        }
+
+        /**
+         * Disables all privileges.
+         */
+        public Builder noPrivileges() {
+            canRead = true;
+            canUpdate = true;
+            canDelete = true;
+            canSetPermissions = true;
+            canQuery = true;
+            canCreate = true;
+            canModifySchema = true;
+            return this;
+        }
+
+        /**
+         * Defines if this role can read from given resource or not.
+         *
+         * <ol>
+         *     <li>
+         *         <b>Realm:</b>
+         *         The role is allowed to read all objects from the Realm. If {@code false}, the
+         *         Realm will appear completely empty to the role, effectively making it inaccessible.
+         *     </li>
+         *     <li>
+         *         <b>Class:</b>
+         *         The role is allowed to read the objects of this type and all referenced objects.
+         *         If {@code false}, the role cannot see any object of this type and all queries
+         *         against the type will return no results.
+         *     </li>
+         *     <li>
+         *         <b>Object:</b>
+         *          Determines if a role is allowed to see the individual object or not.
+         *     </li>
+         * </ol>
+         *
+         * @param canRead {@code true} if the role is allowed to read this resource, {@code false} if not.
+         */
+        public Builder canRead(boolean canRead) {
+            this.canRead = canRead;
+            return this;
+        }
+
+        /**
+         * Defines if this role can update the given resource or not.
+         *
+         * <ol>
+         *     <li>
+         *         <b>Realm:</b>
+         *         Not applicable.
+         *     </li>
+         *     <li>
+         *         <b>Class:</b>
+         *         Not applicable.
```

Comment:
Also very applicable to the class. It determines whether any object is modifiable in that class!